src/subject.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+trait Subject {
+    fn is_valid(&self) -> bool;
+    fn tokens(&self) -> std::io::Result<Vec<&str>>;
```

Comment:
This forces implementors to always allocate a `Vec`. The benefit of a custom `Iterator` is that no additional allocation is required. While iterating through a subject my implementation only copies around some pointers.

Reply:
Very true. And if we made a more concrete type that would be the way to go I think.